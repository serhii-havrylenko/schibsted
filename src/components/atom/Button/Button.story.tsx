import { action } from '@storybook/addon-actions';
import { storiesOf } from '@storybook/react';
import * as React from 'react';

import Button from './Button';

storiesOf('Button', module).add('default', () => (
  <Button onClick={action('clicked')} />
));
