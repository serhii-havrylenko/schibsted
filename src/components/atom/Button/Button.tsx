/**
 * Provides basic button
 * @module #atom/Button
 */
import Btn from '@material-ui/core/Button';
import * as React from 'react';

export interface Button {
  onClick?: () => void;
}

/**
 * Provides basic button component
 */
const Button: React.SFC<Button> = ({ children, onClick }) => (
  <Btn color="primary" onClick={onClick}>
    {children || 'sample'}
  </Btn>
);

export default Button;
