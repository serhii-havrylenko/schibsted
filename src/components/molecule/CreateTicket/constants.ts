export const RequiredError = 'This field is required';
