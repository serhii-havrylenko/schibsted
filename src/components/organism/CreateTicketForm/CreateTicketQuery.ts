import { Mutation } from 'react-apollo';

import {
  CreateTicketQueryData,
  CreateTicketQueryVariables,
} from '#components/types';

export class CreateTicketQuery extends Mutation<
  CreateTicketQueryData,
  CreateTicketQueryVariables
> {}
