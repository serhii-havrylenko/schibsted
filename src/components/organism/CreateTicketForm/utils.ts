import { MutationFn } from 'react-apollo';

import {
  CreateTicketQueryData,
  CreateTicketQueryVariables,
} from '#components/types';

export const createHandler = (
  updateFn: MutationFn<CreateTicketQueryData, CreateTicketQueryVariables>,
) => ({ title, description }: CreateTicketQueryVariables) => {
  updateFn({
    variables: {
      title,
      description,
    },
  });
};
