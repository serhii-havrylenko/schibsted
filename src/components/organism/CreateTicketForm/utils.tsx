import { getOr } from 'lodash/fp';

export const getValueFromRef = (getOr(null, 'current.value') as unknown) as (
  ref: React.RefObject<HTMLInputElement>,
) => string | null;
