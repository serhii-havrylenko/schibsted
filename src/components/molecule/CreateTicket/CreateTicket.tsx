import Card from '@material-ui/core/Card';
import CardActions from '@material-ui/core/CardActions';
import TextField from '@material-ui/core/TextField';
import * as React from 'react';

import Button from '#atom/Button';
import CardContentWrapper from '#atom/CardContentWrapper';
import { RequiredError } from './constants';

export interface CreateTicketProps {
  titleRef: React.RefObject<HTMLInputElement>;
  descriptionRef: React.RefObject<HTMLInputElement>;
  titleError?: boolean;
  descriptionError?: boolean;
  onCreateClick: () => void;
}

const CreateTicket: React.SFC<CreateTicketProps> = ({
  onCreateClick,
  titleRef,
  titleError,
  descriptionRef,
  descriptionError,
}) => (
  <Card>
    <CardContentWrapper>
      <TextField
        margin="dense"
        inputRef={titleRef}
        label="Title"
        required={true}
        type="text"
        error={titleError}
        helperText={titleError && RequiredError}
      />
      <TextField
        margin="dense"
        inputRef={descriptionRef}
        label="Description"
        multiline={true}
        required={true}
        rows={5}
        type="text"
        error={descriptionError}
        helperText={descriptionError && RequiredError}
      />
    </CardContentWrapper>
    <CardActions>
      <Button onClick={onCreateClick}>Create</Button>
    </CardActions>
  </Card>
);

export default CreateTicket;
