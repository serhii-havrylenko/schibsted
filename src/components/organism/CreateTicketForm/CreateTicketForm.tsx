import * as React from 'react';

import CreateTicket from '#molecule/CreateTicket';
import { CREATE_TICKET_QUERY } from '#queries';

import { CreateTicketQueryVariables } from '#components/types';
import { getValueFromRef, updateCacheWithNew } from '#components/utils';
import { CreateTicketQuery } from './CreateTicketQuery';
import { createHandler } from './utils';

export interface CreateTicketFormProps {
  onCreateClick: (payload: CreateTicketQueryVariables) => void;
}

export interface CreateTicketFormState {
  titleError: boolean;
  descriptionError: boolean;
}

export class CreateTicketForm extends React.Component<
  CreateTicketFormProps,
  CreateTicketFormState
> {
  public state = {
    titleError: false,
    descriptionError: false,
  };
  private titleRef: React.RefObject<HTMLInputElement>;
  private descriptionRef: React.RefObject<HTMLInputElement>;

  constructor(props: CreateTicketFormProps) {
    super(props);
    this.titleRef = React.createRef();
    this.descriptionRef = React.createRef();
  }

  public render() {
    const { titleError, descriptionError } = this.state;

    return (
      <CreateTicket
        titleRef={this.titleRef}
        descriptionRef={this.descriptionRef}
        onCreateClick={this.onCreate}
        titleError={titleError}
        descriptionError={descriptionError}
      />
    );
  }

  private onCreate = () => {
    const title = getValueFromRef(this.titleRef);
    const description = getValueFromRef(this.descriptionRef);

    this.setState({
      titleError: !title,
      descriptionError: !description,
    });

    if (!title || !description) {
      return;
    }

    this.props.onCreateClick({ title, description });
  };
}

export default () => (
  <CreateTicketQuery mutation={CREATE_TICKET_QUERY} update={updateCacheWithNew}>
    {(create) => <CreateTicketForm onCreateClick={createHandler(create)} />}
  </CreateTicketQuery>
);
