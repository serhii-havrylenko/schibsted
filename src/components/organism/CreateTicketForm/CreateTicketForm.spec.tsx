import { mount, shallow } from 'enzyme';
import * as React from 'react';

import Button from '#atom/Button';
import CreateTicket from '#molecule/CreateTicket';
import { CreateTicketForm, CreateTicketFormProps } from './CreateTicketForm';

describe('<CreateTicketForm /> tests', () => {
  const props: CreateTicketFormProps = {
    onCreateClick: jest.fn(),
  };

  test('should match snapshot without errors', () => {
    const wrapper = shallow(<CreateTicketForm {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  test('should call onCreateClick with proper params when create clicked', () => {
    const clickMock = jest.fn();
    const wrapper = mount(<CreateTicketForm onCreateClick={clickMock} />);

    ((wrapper
      .find('input')
      .first()
      .instance() as unknown) as HTMLInputElement).value = 'title';

    ((wrapper
      .find('textarea')
      .first()
      .instance() as unknown) as HTMLTextAreaElement).value = 'desc';

    wrapper
      .find(Button)
      .first()
      .simulate('click');

    expect(clickMock).toHaveBeenCalledTimes(1);
    expect(clickMock).toHaveBeenCalledWith({
      title: 'title',
      description: 'desc',
    });
  });

  test('should not call onCreateClick when has invalid fields', () => {
    const clickMock = jest.fn();
    const wrapper = mount(<CreateTicketForm onCreateClick={clickMock} />);

    wrapper
      .find(Button)
      .first()
      .simulate('click');

    expect(clickMock).toHaveBeenCalledTimes(0);
  });

  test('should pass error when no title specified', () => {
    const clickMock = jest.fn();
    const wrapper = mount(<CreateTicketForm onCreateClick={clickMock} />);

    ((wrapper
      .find('textarea')
      .first()
      .instance() as unknown) as HTMLTextAreaElement).value = 'desc';

    wrapper
      .find(Button)
      .first()
      .simulate('click');

    expect(
      wrapper
        .find(CreateTicket)
        .first()
        .props(),
    ).toMatchObject({ titleError: true, descriptionError: false });
  });

  test('should pass error when no description specified', () => {
    const clickMock = jest.fn();
    const wrapper = mount(<CreateTicketForm onCreateClick={clickMock} />);

    ((wrapper
      .find('input')
      .first()
      .instance() as unknown) as HTMLInputElement).value = 'title';

    wrapper
      .find(Button)
      .first()
      .simulate('click');

    expect(
      wrapper
        .find(CreateTicket)
        .first()
        .props(),
    ).toMatchObject({ titleError: false, descriptionError: true });
  });
});
