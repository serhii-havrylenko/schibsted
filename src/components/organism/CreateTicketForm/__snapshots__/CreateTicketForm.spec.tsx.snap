// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateTicketForm /> tests should match snapshot without errors 1`] = `
<CreateTicket
  descriptionError={false}
  descriptionRef={
    Object {
      "current": null,
    }
  }
  onCreateClick={[Function]}
  titleError={false}
  titleRef={
    Object {
      "current": null,
    }
  }
/>
`;
